lib/SILOptimizer/Mandatory/MoveOnlyAddressCheckerUtils.cpp

```diff
@@ -1251,6 +1307,57 @@ struct CopiedLoadBorrowEliminationVisitor : public AccessUseVisitor {
 
 } // namespace
 
+//===----------------------------------------------------------------------===//
+//                  MARK: DestructureThroughDeinit Checking
+//===----------------------------------------------------------------------===//
+
+static void
+checkForDestructureThroughDeinit(MarkMustCheckInst *rootAddress, Operand *use,
+                                 TypeTreeLeafTypeRange usedBits,
+                                 DiagnosticEmitter &diagnosticEmitter) {
+  LLVM_DEBUG(llvm::dbgs() << "    DestructureNeedingUse: " << *use->getUser());
+
+  SILFunction *fn = rootAddress->getFunction();
+  SILModule &mod = fn->getModule();
+
+  // We walk down from our ancestor to our projection, emitting an error if any
+  // of our types have a deinit.
+  TypeOffsetSizePair pair(usedBits);
+  auto targetType = use->get()->getType();
+  auto iterType = rootAddress->getType();
+  TypeOffsetSizePair iterPair(iterType, fn);
+
+  while (iterType != targetType) {
+    // If we have a nominal type as our parent type, see if it has a
+    // deinit. We know that it must be non-copyable since copyable types
+    // cannot contain non-copyable types and that our parent root type must be
+    // an enum, tuple, or struct.
+    if (auto *nom = iterType.getNominalOrBoundGenericNominal()) {
+      if (mod.lookUpMoveOnlyDeinitFunction(nom)) {
```

Comment:
Would it be better to check `nom->getValueTypeDestructor()`, in case the type is from another module and the SIL-level deinit wasn't deserialized?

Reply:
@jckarter I can change that in a follow on commit. I'll also add a note saying deinit declared here.